tests/credentials_provider_sts_web_identity_tests.c

```diff
@@ -0,0 +1,1252 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/testing/aws_test_harness.h>
+
+#include "shared_credentials_test_definitions.h"
+#include <aws/auth/credentials.h>
+#include <aws/auth/private/credentials_utils.h>
+#include <aws/common/clock.h>
+#include <aws/common/condition_variable.h>
+#include <aws/common/date_time.h>
+#include <aws/common/environment.h>
+#include <aws/common/string.h>
+#include <aws/common/thread.h>
+#include <aws/http/request_response.h>
+#include <aws/http/status_code.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/event_loop.h>
+#include <aws/io/logging.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+
+struct aws_mock_sts_web_identity_tester {
+    struct aws_byte_buf request_uri;
+
+    struct aws_array_list response_data_callbacks;
+    bool is_connection_acquire_successful;
+    bool is_request_successful;
+
+    struct aws_mutex lock;
+    struct aws_condition_variable signal;
+
+    struct aws_credentials *credentials;
+    bool has_received_credentials_callback;
+    bool has_received_shutdown_callback;
+
+    int attempts;
+    int response_code;
+};
+
+/**
+ * Redefinition of impl struct because we are mocking connection manager,
+ * whereas free of ctx and connection_options happen in connection
+ * manager's shutdown callback function. We need to free them in our test
+ * manually.
+ */
+struct aws_credentials_provider_sts_web_identity_impl {
+    struct aws_http_connection_manager *connection_manager;
+    struct aws_credentials_provider_system_vtable *function_table;
+    struct aws_string *role_arn;
+    struct aws_string *role_session_name;
+    struct aws_string *token_file_path;
+    struct aws_tls_ctx *ctx;
+    struct aws_tls_connection_options connection_options;
+};
+
+static struct aws_mock_sts_web_identity_tester s_tester;
```

Comment:
nitpick: Since there's only one of these, you can combine the variable decl with the struct decl.

Reply:
yes